### PR TITLE
Add l3 test plan

### PR DESCRIPTION
**Pull request by richardyu-ms**
Add l3 test plan

### CODE REVIEW

File: doc/sai-ptf/L3_test_plan.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+1. run steps 1-10 in test_ecmp_lpm
+2. add next hop for ip ``10.1.2.101`` to the next-hop group in test_ecmp_lpm
+3. Generate Packets, with different source ip ``SIP:192.168.0.1-192.168.0.10`` 
+4. Change other elements in the packets, including ``DIP:192.168.60.1`` ``L4_port``
+5. Verify Packets can be received on LAG1 LAG2 and LAG3, with ``SMAC: SWITCH_MAC_2``
+
+- test_lag_ecmp_lpm_disable
+
+1. run steps 1-8 in test_ecmp_lpm_add
+2. Disable LAG3 members
```
Comment (StormLiangMS) [inline doc/sai-ptf/L3_test_plan.md:273]: how to disable lag3, disable the port or remove it from next hop group?

File: doc/sai-ptf/L3_test_plan.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+- test_ingress_mac_update
+
+1. Generate Packets, with ``SIP:192.168.0.1`` ``DIP:10.1.1.101`` ``DMAC:SWITCH_MAC``
+2. Send packet on Port5
+3. Verify packet received on one of the LAG1's member
+4. Set RIF mac to ``MacX``, the RIF related to Port5
+5. Send packet on Port5 
+6. Verify no packet was received on any LAG1 member
+
+- test_ingress_mtu
```
Comment (StormLiangMS) [inline doc/sai-ptf/L3_test_plan.md:324]: not sure what's the MTU on T0, I heard we can support jumbo frame, should we remove this case or lower the priority?

File: doc/sai-ptf/L3_test_plan.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+- test_trap_cpu
+
+1. Create a Route to ``CPU_PORT`` for ``IP:10.1.100.101``
+2. Create host interface trap on queue4 for IP2ME
+3. send on packet with ``SIP:192.168.0.1`` ``DIP:10.1.100.101`` ``DMAC:SWITCH_MAC`` on port5
+4. verify queue4 increased by 1
+
+- test_rewrite_mac
+
+1. Create neighbor on LAG1 with ``10.1.1.101`` and a new ``DMACX``
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md]: this test is used to check the mac_address attribute of the route object.

File: doc/sai-ptf/L3_test_plan.md (hunk 2 above)
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:324]: this is also for the attribute, can be lower

File: doc/sai-ptf/L3_test_plan.md (hunk 1 above)
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:273]: a member attribute, added more notes

File: doc/sai-ptf/L3_test_plan.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+1. Generate Packets, with ``SIP:192.168.0.1`` ``DIP:192.168.60.1``
+2. Disable the ingress for Port5 - Port8 related RIF
+3. Send packet from Port5 - Port8
+4. Verify no packet was received on any LAG member
+
+
+- test_lag_ecmp_lpm
+
+1. Generate Packets, with different source ip as ``SIP:192.168.0.1-192.168.0.10``
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:253]: yes, the route with a different prefix has already been defined in the config, here we just add a new one.

File: doc/sai-ptf/L3_test_plan.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+1. Set Route on DEST IP ``192.168.11.0/24`` with packet action as SAI_PACKET_ACTION_DROP
+3. Send packet with ``SIP 192.168.0.1`` ``DIP:192.168.11.1`` ``DMAC: SWITCH_MAC``  on port5
+4. verify no packet on any of the LAG1 members
+5. Set Route packet action as SAI_PACKET_ACTION_FORWARD
+6. Send packet with ``SIP 192.168.0.1`` ``DIP:192.168.11.1`` ``DMAC: SWITCH_MAC``  on port5
+7. verify packet received with ``SMAC: SWITCH_MAC`` ``SIP 192.168.0.1`` ``DIP:192.168.11.1`` on one of LAG1 member
+
+- test_route_rif
+
+1. Create route interface for LAG1:``rifx``
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:115]: just check the route can work with RIF, a verify basic check, in actually, it is the basic function for other cases, just isolate the checkpoint here.

File: doc/sai-ptf/L3_test_plan.md (hunk 2 above)
Comment (StormLiangMS) [reply doc/sai-ptf/L3_test_plan.md:325]: I see the MTU setting is 9100 for T0

File: doc/sai-ptf/L3_test_plan.md [hunk 6]
```diff
@@ -0,0 +1,650 @@
...
+1. Send Packets, with ``SIP:192.168.0.1`` ``DIP:192.168.60.1`` 
+2. Verify the packet received on one of the member ports in LAG1-4 (check loadbalanced in LAG and ECMP)
+3. Check and change ecmp_default_hash_seed
+4. Send the same packet as the Step2
+5. Verify the packet received on one of the member ports in LAG1-4 but different from step2 (check loadbalanced in LAG and ECMP)
+
+- test_ingress_no_diff
+
+1. Generate Packets, with ``SIP:192.168.0.1`` ``DIP:192.168.60.1``
+2. Send packet from Port5 - Port8
```
Comment (StormLiangMS) [inline doc/sai-ptf/L3_test_plan.md:498]: Send multi packets?

**Review by StormLiangMS — COMMENTED** [review]
@richardyu-ms add some comments

File: doc/sai-ptf/L3_test_plan.md [hunk 7]
```diff
@@ -0,0 +1,650 @@
...
+7. Send packet with ``DMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.200`` on port5
+8. Received packet with the ``DMAC2``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.200`` on port2
+
+- test_svi_route_flooding
+
+1. Check config for mac for port1~8 already exist
+2. Check already created route and neighbor as, route subnet ``192.168.1.0/24`` next hop to ``VLAN10`` SVI, neighbor for Port2-4 with IP ``DIP:192.168.1.92-94`` bind to ``VLAN10`` SVI
+3. Create neighbor for ``DIP:192.168.1.94/32`` with new ``DMAC:Mac4`` on port4 route interface 
+4. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:00:01:01:99:01:92`` and ``DIP: IP:192.168.1.94`` on port2
+5. No packet or flooding on VLAN10 member port(expect port2)
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:228]: for different ASIC the behavior is not consistent ...

File: doc/sai-ptf/L3_test_plan.md [hunk 8]
```diff
@@ -0,0 +1,650 @@
...
+
+- test_svi_route_learning
+
+1. Run step 1-5 in test_svi_route_flooding
+2. No packet or flooding on VLAN10 member port(expect port2)
+3. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:Mac4`` and ``DIP: IP:192.168.1.92`` on port4 (learn ``Mac4`` on port4)
+4. Received packet with the ``DMAC:00:01:01:99:01:92``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.92`` on port2
+5. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:00:01:01:99:01:92`` and ``DIP: IP:192.168.1.94`` on port2
+6.  Received packet with the ``DMAC:Mac4``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.94`` on port4
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:237]: mac table here

File: doc/sai-ptf/L3_test_plan.md [hunk 9]
```diff
@@ -0,0 +1,650 @@
...
+
+- test_svi_mac_age_after_move
+
+1. Set FDB aging time=10
+2. Run step 1-6 in test_svi_route_learning
+3. wait for FDB aging time
+4. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:Mac4`` and ``DIP: IP:192.168.1.92`` on port4 (learn ``Mac4`` on port4)
+5. Received packet with the ``DMAC:00:01:01:99:01:92``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.92`` on port2
+6. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:00:01:01:99:01:92`` and ``DIP: IP:192.168.1.94`` on port2
+7. Received packet with the ``DMAC:Mac4``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.94`` on port4 (check learnt)
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:273]: Yes, fixed

File: doc/sai-ptf/L3_test_plan.md [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+2. No packet or flooding on VLAN10 member port(expect port2)
+3. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:Mac4`` and ``DIP: IP:192.168.1.92`` on port4 (learn ``Mac4`` on port4)
+4. Received packet with the ``DMAC:00:01:01:99:01:92``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.92`` on port2
+5. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:00:01:01:99:01:92`` and ``DIP: IP:192.168.1.94`` on port2
+6.  Received packet with the ``DMAC:Mac4``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.94`` on port4
+
+- test_svi_mac_move
+
+1. Run step 1-6 in test_svi_route_learning
+2. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:Mac4`` and ``DIP: IP:192.168.1.92`` on port3
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:242]: mac table learn

File: doc/sai-ptf/L3_test_plan.md [hunk 11]
```diff
@@ -0,0 +1,650 @@
...
+- test_drop_route:          Verify drop the packet when SAI_PACKET_ACTION_DROP 
+- test_route_update:        Verify route action gets updated when set attribute from SAI_PACKET_ACTION_DROP to SAI_PACKET_ACTION_FORWARD
+- test_route_lpm_route_nexthop:   Verify lpm route path (with next-hop), route path will be alter to the more accurate one. 
+- test_route_lpm_route_rif:       Verify lpm route path (with rif as next hop), route path will be alter to the more accurate one. 
+- test_svi_route_l3:        Verify route to svi route(through next hop and neighbor)
+- test_svi_route_learning:  Verify route to svi and mac learning
+- test_svi_route_flooding:  Verify route to svi and flooding
+- test_svi_mac_move:  Verify route to svi and mac move
+- test_svi_mac_move_stress: Verify route to svi and mac move (change simultaneously with 100 times in differernt threads)
+- test_default_route: Verify default routr drop action
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:158]: fixed

File: doc/sai-ptf/L3_test_plan.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:237]: change the step for l3 context

File: doc/sai-ptf/L3_test_plan.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:242]: change the step for l3 context

File: doc/sai-ptf/L3_test_plan.md [hunk 12]
```diff
@@ -0,0 +1,650 @@
...
+- test_default_route
+
+1. Make sure default route and route interface are created as config spec.
+2. Send packet with a DIP which is not exist in the config spec for any route or host neighbor, ``IPX``
+3. No packet received
+
+- test_route_diff_prefix_add_then_delete_longer
+
+1. Make sure the route for ``DIP:192.168.1.0/24``is already configured (the route through the next hop)
+2. Add a neighbor with ``IP:192.168.1.0/24`` on Port2 with ``DMAC1``
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:307]: fixed

File: doc/sai-ptf/L3_test_plan.md [hunk 13]
```diff
@@ -0,0 +1,650 @@
...
+3. Send packet with ``DMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.200`` on port5
+4. Received packet with the ``DMAC1``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.200`` on port2
+5. Add a neighbor with ``IP:192.168.1.0/32`` on Port2 with ``DMAC2``
+6. Add a new route for ``DIP: IP:192.168.1.200/32`` with next-hop on port5
+7. Send packet with ``DMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.200`` on port5
+8. Received packet with the ``DMAC2``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.200`` on port2
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:313]: fixed

File: doc/sai-ptf/L3_test_plan.md [hunk 14]
```diff
@@ -0,0 +1,650 @@
...
+6. Send packet for ``DIP:10.1.1.10`` ``DMAC: SWITCH_MAC``  on port5
+7. Verify no packet on any port
+8. Verify the CPU queue0 get one more item
+9. Add the neighbor for ``IP:10.1.1.10`` on LAG1 again
+10. Send packet for ``DIP:10.1.1.10`` ``DMAC: SWITCH_MAC``  on port5
+10. verify packet received with ``SMAC: SWITCH_MAC`` ``SIP 192.168.0.1`` ``DIP:10.1.1.10`` on one of LAG1 member
+
+- test_neighbor_diff_prefix_add_then_remove_longer
+
+1. Add the neighbor for ``IP:10.1.1.10/12`` on LAG1
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:390]: fixed

File: doc/sai-ptf/L3_test_plan.md (hunk 6 above)
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:498]: yes.

File: doc/sai-ptf/L3_test_plan.md [hunk 15]
```diff
@@ -0,0 +1,620 @@
...
+1. Run step 1-6 in test_svi_route_learning
+2. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:ServerX_MAC`` and ``DIP: Port10 Server_IP`` on port3 (mac ``Server_MAC`` move to port3)
+3. Received packet with the ``DMAC:Port10 Server_MAC``, ``SMAC: SWITCH_MAC`` and ``DIP: Port10 Server_IP`` on port10
+4. For check mac learning from L2, send packet with ``DMAC: ServerX_MAC``, ``SMAC:Port2 Server_MAC``  on port2
+5. Received packet with the ``DMAC: ServerX_MAC``, ``SMAC:Port2 Server_MAC`` on port3
+
+- test_svi_mac_move_stress
+
+1. Run step 1-6 in test_svi_route_learning
+2. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:ServerX_MAC`` and ``DIP: Port10 Server_IP`` on port3 (mac ``Server_MAC`` move to port3)
```
Comment (StormLiangMS) [inline doc/sai-ptf/L3_test_plan.md:246]: the SMAC:ServerX_MAC, should be learned ServerX_MAC to port3 other than Server_MAC?

File: doc/sai-ptf/L3_test_plan.md [hunk 16]
```diff
@@ -0,0 +1,620 @@
...
+5. Add a new route for ``DIP: IP:192.168.1.200/32`` and bind to port2 ``rif2`` directly
+6. Add a neighbor with ``IP:192.168.1.200`` on Port1 with ``DMAC2``
+7. Send packet with ``DMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.200`` on port5
+8. Received packet with the ``DMAC2``, ``SMAC: SWITCH_MAC`` and ``DIP: IP:192.168.1.200`` on port1
+
+- test_svi_mac_flooding
+
+1. Check config for mac for port1~8 already exist
+2. Check already created route and neighbor as, route subnet ``192.168.1.0/24`` next hop to ``VLAN10`` SVI, neighbor for Port2-4 with IP ``DIP:192.168.1.92-94`` bind to ``VLAN10`` SVI
+3. Create neighbor for ``DIP:192.168.1.94`` with new ``DMAC:Mac4`` on port4 route interface
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:227]: the new added neighbor should be binded to the vlan rif, then cause this new neighbor mac (target mac) doesn't in arp or mac table, the flood or drop(depends on impl) will happened

File: doc/sai-ptf/L3_test_plan.md [hunk 17]
```diff
@@ -0,0 +1,620 @@
...
+3. Create neighbor for ``DIP:192.168.1.94`` with new ``DMAC:Mac4`` on port4 route interface 
+4. Send packet with ``DMAC: SWITCH_MAC``, ``SMAC:00:01:01:99:01:92`` and ``DIP: IP:192.168.1.94`` on port10
+5. No packet or flooding on VLAN10 member port
+
+- test_svi_mac_learning
+
+1. Run step 1-5 in test_svi_route_flooding
+2. No packet or flooding on VLAN10 member port(expect port2)
+3. For mac learning, send packet with ``SMAC:ServerX_MAC`` ``DMAC: SWITCH_MAC`` and ``DIP: Port10 Server_IP`` on port4 (learn ``Server_MAC`` on port4)
+4. Received packet with the ``DMAC:Port10 Server_MAC``, ``SMAC: SWITCH_MAC`` and ``DIP: Port10 Server_IP`` on port10
```
[the remark this answers is not in the document]
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:236]: when go through port4, the new mac will be learned on the egress port.

File: doc/sai-ptf/L3_test_plan.md (hunk 15 above)
Comment (richardyu-ms) [reply doc/sai-ptf/L3_test_plan.md:251]: Yes

**Review by StormLiangMS — APPROVED** [review]
LGTM

**Pull request merged**